src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -320,24 +336,46 @@ private void EventLoop()
                     // The native shim is responsible for ensuring this condition.
                     Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
 
+                    bool enqueuedEvent = false;
                     for (int i = 0; i < numEvents; i++)
                     {
-                        IntPtr handle = _buffer[i].Data;
+                        IntPtr handle = buffer[i].Data;
                         if (handle == ShutdownHandle)
                         {
                             shutdown = true;
                         }
                         else
                         {
                             Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
-                            _handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
+                            handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
                             if (context != null)
                             {
-                                context.HandleEvents(_buffer[i].Events);
+                                Interop.Sys.SocketEvents events = buffer[i].Events;
+                                events = context.HandleSyncEventsSpeculatively(events);
+                                if (events != Interop.Sys.SocketEvents.None)
+                                {
+                                    var ev = new SocketIOEvent(context, events);
+                                    eventQueue.Enqueue(ev);
+                                    enqueuedEvent = true;
+
+                                    // This is necessary when the JIT generates unoptimized code (debug builds, live debugging,
+                                    // quick JIT, etc.) to ensure that the context does not remain referenced by this method, as
+                                    // such code may keep the stack location live for longer than necessary
+                                    ev = default;
```

Comment:
Does it get optimized away in optimized builds?

Reply:
I didn't check that, it would be fairly easy to optimize this out so I would expect that it would get optimized out. In optimized builds the lifetime seems to reflect at least the scope of the variable, so there's no issue there even without clearing the locals.